### PR TITLE
feat(async-api): Add cancellation token for TreeWalker and RuleRunner

### PR DESCRIPTION
**Pull request by sfoslund**
#### Details

Adding cancellation logic to two of the most time intensive operations performed during scanning, the tree walker and rule runner. 

Based on docs/ discussion online, it appears that the norm is to use `ThrowIfCancellationRequested` to check the cancellation token throughout the async call. This throws a `OperationCanceledException`, which will be wrapped in an `AggregateException` once these methods are wrapped in a task.

Also added bare minimum test coverage for cancellation in TreeWalkerForTest, which previously did not have any unit tests. 

##### Motivation

Part of [1986365](https://mseng.visualstudio.com/1ES/_workitems/edit/1986365) to enable cancellation

##### Context

This is the first in what will be multiple PRs to add cancellation tokens. In the future we will need to add the logic to pass the cancellation token to the methods changed in this PR, but that is blocked until we add a top level async scan API. .

#### Pull request checklist

- [x] Contributes to [1986365](https://mseng.visualstudio.com/1ES/_workitems/edit/1986365)

### CODE REVIEW

File: src/Desktop/UIAutomation/TreeWalkers/TreeWalkerForTest.cs [hunk 1]
```diff
@@ -57,15 +58,17 @@ public TreeWalkerForTest(A11yElement element, BoundedCounter elementCounter)
         /// Refresh tree node data with all children at once.
         /// <param name="mode">indicate the mode</param>
         /// </summary>
-        public void RefreshTreeData(TreeViewMode mode)
+        public void RefreshTreeData(TreeViewMode mode, CancellationToken? cancellationToken = null)
```
Comment (DaveTryon) [inline src/Desktop/UIAutomation/TreeWalkers/TreeWalkerForTest.cs:61]: How do you feel about using `CancellationToken.None` instead of a nullable object?

Comment (sfoslund) [reply src/Desktop/UIAutomation/TreeWalkers/TreeWalkerForTest.cs:61]: Good idea! I'll make that change.

File: src/DesktopTests/UIAutomation/TreeWalkerForTestTests.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (DaveTryon) [inline src/DesktopTests/UIAutomation/TreeWalkerForTestTests.cs:1]: For consistency with the production code, can we move this into the DesktopTests/UIAutomation/TreeWalkers folder and the Axe.Windows.DesktopTests.UIAutomation.TreeWalkers namespace?

File: src/DesktopTests/UIAutomation/TreeWalkerForTestTests.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        [TestMethod]
+        [Timeout(1000)]
+        public void RefreshTreeData_ThrowsWhenCancelled()
+        {
+            var cancellationToken = new CancellationTokenSource();
+            cancellationToken.Cancel();
+            var dataContext = new TreeWalkerDataContext(cancellationToken.Token);
+            var exception = Assert.ThrowsException<OperationCanceledException>(() => _treeWalker.RefreshTreeData(TreeViewMode.Raw, dataContext));
```
Comment (DaveTryon) [inline src/DesktopTests/UIAutomation/TreeWalkerForTestTests.cs:38]: We can simplify this if we aren't actually checking anything on the Exception:
```suggestion
            Assert.ThrowsException<OperationCanceledException>(() => _treeWalker.RefreshTreeData(TreeViewMode.Raw, dataContext));
```

File: src/Actions/Actions/CaptureAction.cs [hunk 4]
```diff
@@ -117,12 +118,13 @@ internal static void PopulateData(ElementDataContext dc, DataContextMode dcMode,
         {
             dc.TreeMode = tm;
             dc.Mode = dcMode;
+            var dataContext = new TreeWalkerDataContext();
```
Comment (DaveTryon) [inline src/Actions/Actions/CaptureAction.cs:121]: Maybe a TODO to help the future integration?
```suggestion
            var dataContext = new TreeWalkerDataContext(); // TODO: Initialize from parameters
```

File: src/RulesTest/RuleRunnerTests.cs [hunk 5]
```diff
@@ -343,6 +345,20 @@ public void RunAll_ReturnsExpectedResults()
...
+        {
+            var e = new MockA11yElement();
+            var ruleMock = new Mock<IRule>(MockBehavior.Strict);
+            var providerMock = new Mock<IRuleProvider>(MockBehavior.Strict);
+            providerMock.Setup(m => m.All).Returns(() => new IRule[] { ruleMock.Object }).Verifiable();
+
+            var runner = new RuleRunner(providerMock.Object);
+            var cancellationToken = new CancellationTokenSource();
+            cancellationToken.Cancel();
+            var exception = Assert.ThrowsException<OperationCanceledException>(() => runner.RunAll(e, cancellationToken.Token));
```
Comment (DaveTryon) [inline src/RulesTest/RuleRunnerTests.cs:359]: We can simplify this if we don't need to use the exception:
```suggestion
            Assert.ThrowsException<OperationCanceledException>(() => runner.RunAll(e, cancellationToken.Token));
```

**Review by DaveTryon — APPROVED** [review]

**Review by DaveTryon — APPROVED** [review]
LGTM

**Pull request merged**